File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    player_indices = {name: i for i, name in enumerate(players)}
    
    for called_name in callings:
        current_idx = player_indices[called_name]
        
        prev_idx = current_idx - 1
        prev_name = players[prev_idx]
        
        players[prev_idx], players[current_idx] = players[current_idx], players[prev_idx]
        
        player_indices[called_name] = prev_idx
        player_indices[prev_name] = current_idx
        
    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    const playerMap = {};

    for (let i = 0; i < players.length; i++) {
      playerMap[players[i]] = i;
    }
    
    for (let calling of callings) {
        const curIdx = playerMap[calling];
        const frontPlayer = players[curIdx - 1];

        players[curIdx - 1] = calling;
        players[curIdx] = frontPlayer;

        playerMap[calling] = curIdx - 1;
        playerMap[frontPlayer] = curIdx;
    }
    
    return players;
}